pkg/cli/cli.go

```diff
@@ -0,0 +1,304 @@
+	// Run runs the CLI, usually returning an error if command line configuration
+	// is incorrect.
+	Run() error
+}
+
+// Option is a function that can configure the cli
+type Option func(*cli) error
+
+// cli defines the command line structure and interfaces that are used to
+// scaffold kubebuilder project files.
+type cli struct {
+	// Base command name. Can be injected downstream.
+	commandName string
+	// Default project version. Used in CLI flag setup.
+	defaultProjectVersion string
+	// Project version to scaffold.
+	projectVersion string
+	// True if the project has config file.
+	configured bool
+
+	// Base command.
+	cmd *cobra.Command
+	// Commands injected by options.
+	extraCommands []*cobra.Command
+	// Plugins injected by options.
+	plugins map[string][]plugin.Base
+}
+
+// New creates a new cli instance.
+func New(opts ...Option) (CLI, error) {
+	c := &cli{
+		commandName:           "kubebuilder",
+		defaultProjectVersion: internalconfig.DefaultVersion,
+		plugins:               map[string][]plugin.Base{},
+	}
+	for _, opt := range opts {
+		if err := opt(c); err != nil {
+			return nil, err
+		}
+	}
+
+	if err := c.initialize(); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// Run runs the cli.
+func (c cli) Run() error {
+	return c.cmd.Execute()
+}
+
+// WithCommandName is an Option that sets the cli's root command name.
+func WithCommandName(name string) Option {
+	return func(c *cli) error {
+		c.commandName = name
+		return nil
+	}
+}
+
+// WithDefaultProjectVersion is an Option that sets the cli's default project
+// version. Setting an unknown version will result in an error.
+func WithDefaultProjectVersion(version string) Option {
+	return func(c *cli) error {
+		c.defaultProjectVersion = version
+		return nil
+	}
+}
+
+// WithPlugins is an Option that sets the cli's plugins.
+func WithPlugins(plugins ...plugin.Base) Option {
+	return func(c *cli) error {
+		for _, p := range plugins {
+			for _, version := range p.SupportedProjectVersions() {
+				if _, ok := c.plugins[version]; !ok {
+					c.plugins[version] = []plugin.Base{}
+				}
+				c.plugins[version] = append(c.plugins[version], p)
+			}
+		}
+		return nil
+	}
+}
+
+// WithExtraCommands is an Option that adds extra subcommands to the cli.
+// Adding extra commands that duplicate existing commands results in an error.
+func WithExtraCommands(cmds ...*cobra.Command) Option {
+	return func(c *cli) error {
+		c.extraCommands = append(c.extraCommands, cmds...)
+		return nil
+	}
+}
+
+// initialize initializes the cli.
+func (c *cli) initialize() error {
+	// Configure the project version first for plugin retrieval in command
+	// constructors.
+	projectConfig, err := internalconfig.Read()
+	if os.IsNotExist(err) {
+		c.configured = false
+		c.projectVersion = c.defaultProjectVersion
+	} else if err == nil {
+		c.configured = true
+		c.projectVersion = projectConfig.Version
+	} else {
+		return fmt.Errorf("failed to read config: %v", err)
+	}
```

Comment:
Just a nit. Suggestion ( I think make more clear when has an error and not what should be done )

```go
if err != nil {
      if os.IsNotExist(err) {
            c.configured = false
	    c.projectVersion = c.defaultProjectVersion
      } else {
           return fmt.Errorf("failed to read config: %v", err)
      }
} else {
     c.configured = true
     c.projectVersion = projectConfig.Version
}
```

Reply:
This is less clear IMO.